### PR TITLE
Implement sheared toggle button

### PR DESCRIPTION
**Pull request by bdach**
Another piece for the new mod screen (#16917).

https://user-images.githubusercontent.com/20418176/160930258-414cf2cd-4724-47d0-a3af-6cf8c15446da.mp4

It is a toggle button. It supports being toggled and disabled in both toggle states. I intend to use it to toggle the mod customisation panel [like depicted here](https://github.com/ppy/osu/issues/16917#issuecomment-1060022752).

This was not on design, but the visuals were made to [match the other buttons on the design](https://www.figma.com/file/NkdREXr6wFkA99oLw8PnBv/Client%2FMod-Selection?node-id=0%3A1).

As an aside I'm generally not sure if any of the other buttons from design linked above will return.
* The "back" button is weird because the popup has an "X" button in the top right, so not sure there is much point in adding the "back" button at the bottom too.
* The "confirm" button shouldn't be needed because I intend the changes to immediately take effect like they do on the current mod overlay.
* The "mods" button is essentially another "close" option, so see point 1.
* And the other buttons ("random"/"options") would change beatmap/open beatmap options... in the background? which is like ultra weird.

### CODE REVIEW

**Comment by smoogipoo** [comment]
Is there a non-toggle version of this? I believe this is the same button design that flyte wants to use for the back button / "go" button in the toolbar on other screens, e.g.:
image
image

**Comment by bdach** [comment]
I don't believe there is one in master, and I don't have one written myself because it looked unlikely that I'd need it (for reasons mentioned in OP).

**Review by smoogipoo — DISMISSED** [review]
Feels good to me

File: osu.Game/Graphics/UserInterface/ShearedToggleButton.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        private readonly Box background;
+        private readonly OsuSpriteText text;
+
+        private Sample? sampleOff;
+        private Sample? sampleOn;
+
+        [Resolved]
+        private OverlayColourProvider colourProvider { get; set; } = null!;
+
+        public ShearedToggleButton(float shear)
```
Comment (peppy) [inline osu.Game/Graphics/UserInterface/ShearedToggleButton.cs:43]: Are there going to be multiple `shear` amounts in the design? If not, I'd hardcode this.

File: osu.Game/Graphics/UserInterface/ShearedToggleButton.cs [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        {
+            Height = 50;
```
Comment (peppy) [inline osu.Game/Graphics/UserInterface/ShearedToggleButton.cs:45]: I know that this is following by example, but it's very weird that buttons have a hidden requirement of their width being manually set.

How would you feel about something like

```diff
diff --git a/osu.Game/Graphics/UserInterface/ShearedToggleButton.cs b/osu.Game/Graphics/UserInterface/ShearedToggleButton.cs
index acbed29279..c7cb2eb33c 100644
--- a/osu.Game/Graphics/UserInterface/ShearedToggleButton.cs
+++ b/osu.Game/Graphics/UserInterface/ShearedToggleButton.cs
@@ -40,9 +40,10 @@ public LocalisableString Text
         [Resolved]
         private OverlayColourProvider colourProvider { get; set; } = null!;
 
-        public ShearedToggleButton(float shear)
+        public ShearedToggleButton(float shear, float? width = null)
         {
             Height = 50;
+
             Padding = new MarginPadding { Horizontal = shear * 50 };
 
             Content.CornerRadius = 7;
@@ -65,6 +66,16 @@ public ShearedToggleButton(float shear)
                     Shear = new Vector2(-shear, 0)
                 }
             };
+
+            if (width != null)
+            {
+                Width = width.Value;
+            }
+            else
+            {
+                AutoSizeAxes = Axes.X;
+                text.Margin = new MarginPadding { Horizontal = 10 };
+            }
         }
 
         [BackgroundDependencyLoader]

```

I'd probably want the `width` parameter to be first (and have xmldoc and tests), above code is only intended to show intent.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Feel wise this implementation is fine.

File: osu.Game/Graphics/UserInterface/ShearedToggleButton.cs (hunk 1 above)
Comment (bdach) [reply osu.Game/Graphics/UserInterface/ShearedToggleButton.cs:43]: I don't foresee any, was a bit of future-proofing at best. I'll remove for now.

File: osu.Game/Graphics/UserInterface/ShearedToggleButton.cs (hunk 2 above)
Comment (bdach) [reply osu.Game/Graphics/UserInterface/ShearedToggleButton.cs:57]: Sure, have applied (with slight adjustments).

**Review by peppy — APPROVED** [review]

**Comment by hello-smile6** [comment]
Why shear buttons?

**Comment by hello-smile6** [comment]
Are they bunnies in disguise?

**Pull request merged**